share/shwap/p2p/bitswap/bitswap.go

```diff
@@ -0,0 +1,136 @@
+	// multiple requests and only after reads up the data in portions one-by-one targeting the
+	// targetMessageSize.
+	//
+	// Bigger number will speed transfers up if reading data from disk is fast. In our case, the
+	// Bitswap's size lookup via [Blockstore] will already cause underlying cache to keep the data,
+	// so reading up data is fast, and we can aim to pack as much as we can.
+	targetMessageSize = 1 << 20 // 1MB
+	// outstandingBytesPerPeer limits number of bytes queued for work for a peer across multiple requests.
+	// We set it to be equal to targetMessageSize * N, so there can max N messages being prepared for
+	// a peer at once.
+	outstandingBytesPerPeer = targetMessageSize * 4
+)
+
+// NewNetwork constructs Bitswap network for Shwap protocol composition.
+func NewNetwork(host host.Host, prefix protocol.ID) network.BitSwapNetwork {
+	prefix = shwapProtocolID(prefix)
+	net := network.NewFromIpfsHost(
+		host,
+		routinghelpers.Null{},
+		network.Prefix(prefix),
+		network.SupportedProtocols([]protocol.ID{protocolID}),
+	)
+	return net
+}
+
+// NewClient constructs a Bitswap client with parameters optimized for Shwap protocol composition.
+// Meant to be used by Full and Light nodes.
+func NewClient(
+	ctx context.Context,
+	net network.BitSwapNetwork,
+	bstore blockstore.Blockstore,
+) *client.Client {
+	return client.New(
+		ctx,
+		net,
+		bstore,
+		client.SetSimulateDontHavesOnTimeout(simulateDontHaves),
+		client.ProviderSearchDelay(providerSearchDelay),
+		client.RebroadcastDelay(delay.Fixed(rebroadcastDelay)),
+		// Prevents Has calls to Blockstore for metric that counts duplicates
+		// Unnecessary for our use case, so we can save some disk lookups.
+		client.WithoutDuplicatedBlockStats(),
+	)
+}
+
+// NewServer construct a Bitswap server with parameters optimized for Shwap protocol composition.
+// Meant to be used by Full nodes.
+func NewServer(
+	ctx context.Context,
+	net network.BitSwapNetwork,
+	bstore blockstore.Blockstore,
+	opts ...server.Option,
+) *server.Server {
+	opts = append(
+		opts,
+		server.ProvideEnabled(providesEnabled),
+		server.SetSendDontHaves(sendDontHaves),
+		server.MaxQueuedWantlistEntriesPerPeer(maxServerWantListsPerPeer),
+		server.WithTargetMessageSize(targetMessageSize),
+		server.MaxOutstandingBytesPerPeer(outstandingBytesPerPeer),
+	)
+
+	return server.New(
+		ctx,
+		net,
+		bstore,
+		opts...,
+	)
+}
+
+// TODO(@Wondertan): We have to use the protocol defined by Bitswap here
+//
+//	due to a little bug. Bitswap allows setting custom protocols, but
+//	they have to be either one of the switch.
+//	https://github.com/ipfs/boxo/blob/dfd4a53ba828a368cec8d61c3fe12969ac6aa94c/bitswap/network/ipfs_impl.go#L250-L266
+var protocolID = network.ProtocolBitswap
+
+func shwapProtocolID(network protocol.ID) protocol.ID {
+	if network == "" {
+		return ""
+	}
+	return protocol.ID(fmt.Sprintf("/%s/shwap", network))
```

Comment:
Maybe we should add versioning in case protocol breaking changes?

Reply:
Another question I have about protocol ID is the idea we discussed to separate bitswap networks into samples and reconstruction related. Any plans to make constructors aware of network type in this PR? it seems like a PR those changes would fit well

Reply:
I thought of including that in this PR, but decided against that as those constructors should land next to reconstruction retrieval.